fdb-record-layer-spatial/src/main/java/com/geophile/z/async/SpatialJoinIteratorAsync.java

```diff
@@ -0,0 +1,287 @@
+    {
+        final SortedArray<RecordWithSpatialObject> queryIndex = new SortedArray.OfBaseRecord();
+        try {
+            final SpatialIndex<RecordWithSpatialObject> querySpatialIndex = SpatialIndex.newSpatialIndex(dataSpatialIndex.space(),
+                    queryIndex,
+                    querySpatialObject.maxZ() == 1
+                    ? SpatialIndex.Options.SINGLE_CELL
+                    : SpatialIndex.Options.DEFAULT);
+            querySpatialIndex.add(querySpatialObject,
+                    () -> {
+                        RecordWithSpatialObject queryRecord = queryIndex.newRecord();
+                        queryRecord.spatialObject(querySpatialObject);
+                        return queryRecord;
+                    });
+        } catch (IOException | InterruptedException ex) {
+            throw new RuntimeException(ex);
+        }
+        final IndexAsync<LEFT_RECORD> queryIndexAsync = (IndexAsync<LEFT_RECORD>)IndexAsync.fromSync(queryIndex);
+        final SpatialIndexImplAsync<LEFT_RECORD> querySpatialIndexAsync = new SpatialIndexImplAsync<>((SpaceImpl) dataSpatialIndex.space(),
+                queryIndexAsync,
+                querySpatialObject.maxZ() == 1 ? SpatialIndexAsync.Options.SINGLE_CELL : SpatialIndexAsync.Options.DEFAULT);
+        SpatialJoinOutputAsync<LEFT_RECORD, RIGHT_RECORD> pendingLeftRight =
+                (left, right) -> {
+                    if (filter.overlap(querySpatialObject, right)) {
+                        pending.add((T)outputGenerator.generateOutput(left, right));
+                    }
+                };
+        left = SpatialJoinInputAsync.newSpatialJoinInputAsync(querySpatialIndexAsync,
+                                                    pendingLeftRight,
+                                                    leftInputObserver);
+        SpatialJoinOutputAsync<RIGHT_RECORD, LEFT_RECORD> pendingRightLeft =
+                (right, left) -> {
+                    if (filter.overlap(querySpatialObject, right)) {
+                        pending.add((T)outputGenerator.generateOutput(left, right));
+                    }
+                };
+        right = SpatialJoinInputAsync.newSpatialJoinInputAsync(dataSpatialIndex,
+                                                     pendingRightLeft,
+                                                     rightInputObserver);
+        left.otherInput(right);
+        right.otherInput(left);
+        if (LOG.isLoggable(Level.INFO)) {
+            LOG.log(Level.INFO,
+                    "SpatialJoinIterator {0}: {1} x {2}",
+                    new Object[]{this, left, right});
+        }
+    }
+
+    private CompletableFuture<Void> ensurePending()
+    {
+        if (pending.isEmpty()) {
+            return findPairs();
+        } else {
+            return CompletableFutures.NULL;
+        }
+    }
+
+    // The synchronous version also calls this from the constructors. That's hard when it's async, so just don't bother.
```

Comment:
I'm not sure I understand this comment. Is this method not used and/or is it unnecessary?